### PR TITLE
[crypto] Add an AES-GCM GHASH implementation.

### PR DESCRIPTION
**Pull request by jadephilipoom**
This is one of the two core subroutines of AES-GCM (the other being GCTR). It's Algorithm 2 in [NIST SP800-38D](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf).

The most complex operation here is polynomial multiplication (`galois_mul` in the code). In a nutshell, we interpret 128 bits `a0, a1, a2, ...a127` as the polynomial `a0 + a1x + a2x^2 + ... + a127x^127`, where the `a` values must all be 0 or 1. This represents an element in a 128-bit Galois field, where addition/subtraction are XOR and multiplication is computed modulo a degree-128 polynomial.

This implementation uses precomputed tables to speed up polynomial multiplication, as is standard across other implementations. There are two top-level functions: an `aes_gcm_ghash_update` that performs a single step and a one-shot `aes_gcm_ghash` that will be eventually exposed to the crypto library API. The `_update` version will be used internally in encryption to speed up the computation and reduce awkward allocations and buffer-wrangling, but will not be exposed.

Locally, I've actually implemented GCTR and encryption as well, because I couldn't easily generate test data for the smaller subroutines. So this code has been tested, but I'm merging the work piecemeal because otherwise it would be too big.

### CODE REVIEW

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+/**
+ * Reverse the bits of a byte.
+ *
+ * @param word Input word.
+ * @return Word with bytes reversed.
+ */
+static uint8_t reverse_bits(uint8_t byte) {
+  uint8_t out;
+  for (size_t i = 0; i < 8; i++) {
+    out <<= 1;
```
[the remark this answers is not in the document]
Comment (jadephilipoom) [reply sw/device/lib/crypto/aes_gcm/aes_gcm.c:86]: Regardless of whether it's technically UB or not, I intended to set it to 0 -- will fix. Thanks for the catch!

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+    galois_mul(state, H_table, &yH);
+    memcpy(state->data, yH.data, kAesBlockNumBytes);
+  }
+}
+
+aes_error_t aes_gcm_ghash(const aes_block_t *hash_subkey,
+                          const size_t input_len, const uint8_t *input,
+                          aes_block_t *output) {
+  // Compute the product table for H.
+  aes_block_t H_table[256];
```
[the remark this answers is not in the document]
Comment (jadephilipoom) [reply sw/device/lib/crypto/aes_gcm/aes_gcm.c:333]: For an n-bit window, this method requires:
- A reduction table (constant) with `2^n` elements, `8+n` bits each
- A product table (computed once per encryption or decryption operation) with `2^n` elements, 128 bits each

So the total size for this implementation with an 8-bit window is 4.6 kiB. It can pretty straightforwardly be adapted to use a 4-bit window instead, which would be much smaller (about 280 bytes) but slower. Let me actually try it out and benchmark it so we know what the tradeoffs are; I think it will be significantly slower but I'm not certain how much. We could also consider setting things up so that one can choose which setup to use at compile time -- space vs speed.

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+ * is big-endian.
+ *
+ * @param word Input word.
+ * @return Word with bytes reversed.
+ */
+static uint32_t reverse_bytes(uint32_t word) {
+  uint8_t *word_bytes = (uint8_t *)&word;
+  uint32_t out;
+  for (size_t i = 0; i < sizeof(uint32_t); i++) {
+    out <<= 8;
```
Comment (dmcardle) [inline sw/device/lib/crypto/aes_gcm/aes_gcm.c:106]: Another read of uninitialized `out` :)

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+
+/**
+ * Retrieve the byte at a given index from the AES block.
+ *
+ * @param block Input AES block
+ * @param index Index of byte to retrieve (must be < `kAesBlockNumBytes`)
+ * @return Value of block[index]
+ */
+static inline uint8_t block_byte_get(const aes_block_t *block, size_t index) {
+  return ((uint8_t *)block->data)[index];
```
Comment (dmcardle) [inline sw/device/lib/crypto/aes_gcm/aes_gcm.c:146]: To be very pedantic, isn't this a strict aliasing violation since `block->data` is an array of uint32_t and uint8_t is not guaranteed to be a char?

Comment (alphan) [reply sw/device/lib/crypto/aes_gcm/aes_gcm.c:146]: I think so (and I think we might have other instances of this). How about:

```suggestion
  return (uint8_t)((char *)block->data)[index];
```

**Comment by jadephilipoom** [comment]
> Hi @jadephilipoom , I might not know all the involved constraints here but I thought it would be worth mentioning anyway: Ibex in OpenTitan as well as our toolchains have support for the RISC-V Bitmanip extension ([see the Ibex documentation for more details](https://ibex-core.readthedocs.io/en/latest/03_reference/instruction_decode_execute.html#bit-manipulation-extension)), meaning you can leverage the `clmul(h)` and `grev` instructions for improving performance/code size of GHASH as outlined in [this paper (see Section 4)](https://tches.iacr.org/index.php/TCHES/article/view/8729/8329).

Thanks! I'm aware and plan to adjust the implementation to use the bitmanip instructions, but thought it would be helpful to get the simplest variant working first so that we have passing tests to start from :slightly_smiling_face: I was also under the impression that the bitmanip instructions were not quite ready to use, but I guess I was mistaken. Was toolchain support added recently?

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (jadephilipoom) [reply sw/device/lib/crypto/aes_gcm/aes_gcm.c:333]: Okay, here's some rough benchmarks:

| Window size  | Storage (bytes) | Overhead (cycles) | GHASH (cycles) |
| --------- | -------- | ------------- |-------------|
|  4 bits  |   280  | 17,100 | 20,100 |
|  8 bits  |   4600 |  170,300 | 10,700 |

The "overhead" happens once per encryption (and is dominated in the 8-bit case by computing the product table). The GHASH happens 1-2x per encryption (depending on IV length) and then 1x per 128-bit block of plaintext OR aad. So the 4-bit window is faster for smaller messages, but scales much worse:

|  Plaintext + aad size | 4 bits (cycles / encrypt) | 8 bits (cycles / encrypt) |
| ------------- | ----------- | ----------- |
|  0 bytes    |   57,200 | 191,700 |
| 128 bytes | 218,000 | 277,300 |
| 256 bytes | 378,800 | 362,900 |
| 384 bytes | 539,600 | 448,500 |
| 512 bytes | 700,400 | 534,100 |
...
...
...
| 1024 bytes | 1,343,600 | 876,500 |

However, looking at these cycle counts, we're already looking at quite high latency for these input sizes; at 100 MHz, the 512-byte case takes 5 ms even for the 8-bit version. Since we're not offering a streaming API, this is probably pushing the edge of where latency requirements would require chunking the input anyway.

An important note here is that I believe using bitmanip instructions will remove the need for precomputed tables of any size.

Bottom line: Given the relatively late point at which speed benefits appear, it's probably worth prioritizing space and switching to a 4-bit window. Long-term, bitmanip instructions will replace tables.

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c (hunk 4 above)
Comment (jadephilipoom) [reply sw/device/lib/crypto/aes_gcm/aes_gcm.c:146]: I could also change all these bytes to `char` instead. I used `uint8_t` because I was modelling off of the pointer casts in `memory.c`: https://github.com/lowRISC/opentitan/blob/a96565ddbe684f14663c187db8a6c17908eebfbc/sw/device/lib/base/memory.c#L13-L20

I guess passing through `void *` avoids the strict aliasing issue?

Comment (dmcardle) [reply sw/device/lib/crypto/aes_gcm/aes_gcm.c:128]: My understanding is that strict aliasing is the assumption that pointers of different types will never alias the same memory location. There are exceptions that allow for casting to character-type pointers, casting any `T *` to `const T *`, etc.

I don't believe passing through `void *` avoids the potential aliasing issue — the input pointers could have been something other than `uint8_t *`. If we're being very strict, we should use `char *` in the implementation of memcpy() as well.

However, if we're comfortable assuming that `uint8_t` is typedefed as `unsigned char` (and always will be), I guess my original point about strict aliasing is moot. In that case, it's more of a style issue.

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+
+aes_error_t aes_gcm_ghash(const aes_block_t *hash_subkey,
+                          const size_t input_len, const uint8_t *input,
+                          aes_block_t *output) {
+  // Compute the product table for H.
+  aes_block_t H_table[256];
+  make_product_table(hash_subkey, H_table);
+
+  // If the input length is not a multiple of the block size, fail.
+  if (input_len % kAesBlockNumBytes != 0) {
```
[the remark this answers is not in the document]
Comment (alphan) [reply sw/device/lib/crypto/aes_gcm/aes_gcm.c]: I agree but I don't think we depend on `libgcc` anymore so most probably these are optimized. See #11451 for context.

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+
+/**
+ * Precomputed modular reduction constants for Galois field multiplication.
+ *
+ * This implementation uses 4-bit windows. The bytes here represent 12-bit
+ * little-endian values.
+ *
+ * The entry with index i in this table is equal to i * 0xe1, where the bytes i
+ * and 0xe1 are interpreted as polynomials in the GCM Galois field. For
+ * example, 0xe1 = 11100001 is the degree-8 polynomial x^8 + x^2 + x + 1.
```
Comment (alphan) [inline sw/device/lib/crypto/aes_gcm/aes_gcm.c:26]: nit:
```suggestion
 * example, 0xe1 = 0b1110_0001 is the degree-8 polynomial x^8 + x^2 + x + 1.
```

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+static const uint16_t kGFReduceTable[16] = {
+    0x0000, 0x201c, 0x4038, 0x6024, 0x8070, 0xa06c, 0xc048, 0xe054,
+    0x00e1, 0x20fd, 0x40d9, 0x60c5, 0x8091, 0xa08d, 0xc0a9, 0xe0b5};
+
+/**
+ * Performs a bitwise XOR of two blocks.
+ *
+ * This operation corresponds to addition in the Galois field.
+ *
+ * @param condition Either 0 or 1
```
Comment (alphan) [inline sw/device/lib/crypto/aes_gcm/aes_gcm.c:46]:
```suggestion
```

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c (hunk 4 above)
Comment (jadephilipoom) [reply sw/device/lib/crypto/aes_gcm/aes_gcm.c:128]: Okay, I've made the change that Alphan suggested here and also fixed similar places elsewhere in the code. Here's the diff (since I'm going to rebase it into a previous commit, so it won't be straightforward to view here); mind taking a quick look?
```diff
@@ -82,11 +82,10 @@ static uint8_t reverse_bits(uint8_t byte) {
  * @return Word with bytes reversed.
  */
 static uint32_t reverse_bytes(uint32_t word) {
-  uint8_t *word_bytes = (uint8_t *)&word;
   uint32_t out = 0;
   for (size_t i = 0; i < sizeof(uint32_t); i++) {
     out <<= 8;
-    out |= word_bytes[i];
+    out |= (word >> (8 * i)) & 255;
   }
   return out;
 }
@@ -125,7 +124,7 @@ static inline void block_shiftr(aes_block_t *block, size_t nbits) {
  * @return Value of block[index]
  */
 static inline uint8_t block_byte_get(const aes_block_t *block, size_t index) {
-  return ((uint8_t *)block->data)[index];
+  return (uint8_t)((char *)block->data)[index];
 }
 
 /**
@@ -311,9 +310,8 @@ static void aes_gcm_ghash_update(const aes_block_t H_table[16],
     if ((i == nblocks - 1) && (get_block_offset(input_len) != 0)) {
       // Last block is not full; pad with zeroes.
       size_t nbytes = get_block_offset(input_len);
-      uint8_t *input_data = (uint8_t *)input_block.data;
-      memcpy(input_data, input, nbytes);
-      memset(input_data + nbytes, 0, kAesBlockNumBytes - nbytes);
+      memset(input_block.data, 0, kAesBlockNumBytes);
+      memcpy(input_block.data, input, nbytes);
     } else {
       // Full block; copy data and advance input pointer.
       memcpy(input_block.data, input, kAesBlockNumBytes);
```
The last change is slightly less efficient than what I had before, but I don't think it's a big difference and this version is more readable anyway :slightly_smiling_face:

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+  for (size_t i = 0; i < kAesBlockNumWords; i++) {
+    out->data[i] = x->data[i] ^ y->data[i];
+  }
+}
+
+/**
+ * Reverse the bits of a 4-bit number.
+ *
+ * @param word Input word.
+ * @return Word with bytes reversed.
```
Comment (alphan) [inline sw/device/lib/crypto/aes_gcm/aes_gcm.c:63]: nit: this comment should be updated (it looks like this function is reversing bits of a byte, not the bytes of a word)

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+ */
+static uint32_t reverse_bytes(uint32_t word) {
+  uint32_t out = 0;
+  for (size_t i = 0; i < sizeof(uint32_t); i++) {
+    out <<= 8;
+    out |= (word >> (i << 3)) & 255;
+  }
+  return out;
+}
```
Comment (alphan) [inline sw/device/lib/crypto/aes_gcm/aes_gcm.c:91]: Please take a look at `_ot_builtin_bswap_i32` or `__bswapsi2` in `sw/device/lib/base/math_builtins.c` unless you have other plans for this function.

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+  // to the last block's carry, and then reverse the bytes back.
+  uint32_t carry = 0;
+  for (size_t i = 0; i < kAesBlockNumWords; i++) {
+    uint32_t rev = reverse_bytes(block->data[i]);
+    uint32_t next_carry = rev & ((1 << nbits) - 1);
+    rev >>= nbits;
+    rev |= carry << (32 - nbits);
+    carry = next_carry;
+    block->data[i] = reverse_bytes(rev);
+  }
```
Comment (alphan) [inline sw/device/lib/crypto/aes_gcm/aes_gcm.c:116]: nit: can't we do this in the opposite order? This would eliminate two calls to `reverse_bytes()`.

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+ * @return Word with bytes reversed.
+ */
+static uint8_t reverse_bits(uint8_t byte) {
+  uint8_t out = 0;
+  for (size_t i = 0; i < 4; i++) {
+    out <<= 1;
+    out |= (byte >> i) & 1;
+  }
+  return out;
+}
```
Comment (alphan) [inline sw/device/lib/crypto/aes_gcm/aes_gcm.c:72]: How about something like:
```
byte = (byte & 0x55) << 1 | (byte & 0xaa) >> 1;
byte = (byte & 0x33) << 2 | (byte & 0xcc) >> 2;
byte = (byte & 0x0f) << 4 | (byte & 0xf0) >> 4;
return byte;
```
This is the byte version of the first algorithm in Hacker's Delight 7.1, 2nd ed ([link](https://learning.oreilly.com/library/view/hackers-delight-second/9780133084993/ch07.html#:-:text=x%20%3D%20(x%20%26%200x55555555)%C2%A0%C2%A0%3C%3C%C2%A0%C2%A0%C2%A01,16%20%7C%20(x%20%26%200xFFFF0000)%20%3E%3E%2016%3B), also some discussion on [SO](https://stackoverflow.com/a/2602885)).

Side note: we should also consider the bitmanip extensions.

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+ * Construct a product table for the hash subkey H.
+ *
+ * The product table entry at index i is equal to i * H, where both i and H are
+ * interpreted as Galois field polynomials.
+ *
+ * @param H Hash subkey
+ * @param product_table Buffer for output
+ */
+static void make_product_table(const aes_block_t *H,
+                               aes_block_t product_table[16]) {
```
Comment (alphan) [inline sw/device/lib/crypto/aes_gcm/aes_gcm.c:193]: nit: wdyt about defining a type for `aes_block_t product_table[16]`

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c [hunk 13]
```diff
@@ -0,0 +1,351 @@
...
+ * @param state GHASH state, updated in place
+ */
+static void aes_gcm_ghash_update(const aes_block_t H_table[16],
+                                 const size_t input_len, const uint8_t *input,
+                                 aes_block_t *state) {
+  // Calculate the number of AES blocks needed for the input.
+  size_t nblocks = get_nblocks(input_len);
+
+  // Main loop.
+  for (size_t i = 0; i < nblocks; i++) {
```
Comment (alphan) [inline sw/device/lib/crypto/aes_gcm/aes_gcm.c:307]:
```suggestion
  for (size_t i = 0; i < nblocks; ++i) {
```

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c [hunk 14]
```diff
@@ -0,0 +1,351 @@
...
+    galois_mul(state, H_table, &yH);
+    memcpy(state->data, yH.data, kAesBlockNumBytes);
+  }
+}
+
+aes_error_t aes_gcm_ghash(const aes_block_t *hash_subkey,
+                          const size_t input_len, const uint8_t *input,
+                          aes_block_t *output) {
+  // Compute the product table for H.
+  aes_block_t H_table[16];
```
Comment (alphan) [inline sw/device/lib/crypto/aes_gcm/aes_gcm.c:333]: nit: this looks like a large local variable, does it have to be on stack?

**Review by alphan — APPROVED** [review]
Thanks @jadephilipoom!

I'm not familiar with the algorithm so my comments are mostly about efficiency.

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c (hunk 4 above)
Comment (dmcardle) [reply sw/device/lib/crypto/aes_gcm/aes_gcm.c]: > The last change is slightly less efficient than what I had before, but I don't think it's a big difference and this version is more readable anyway 🙂

Okay, I see it used to memcpy to `input_data[0..nbytes]` and then zero out the rest, but now it zeroes out the whole block before the memcpy. Was that motivated by the strict aliasing concern?

Other than that, the inline diff LGTM!

Comment (jadephilipoom) [reply sw/device/lib/crypto/aes_gcm/aes_gcm.c]: > Okay, I see it used to memcpy to `input_data[0..nbytes]` and then zero out the rest, but now it zeroes out the whole block before the memcpy. Was that motivated by the strict aliasing concern?

The previous implementation casted the `uint32_t *` to a `uint8_t *` in order to make the pointer arithmetic work out for the second write (`input_data + nbytes`). But this was a little hacky and I'd been considering setting it all to zero before the copy anyway; the fact that the cast needed to make the pointer arithmetic nice violated strict aliasing just made the style/speed decision easier!

**Comment by alphan** [comment]
> > Hi @jadephilipoom , I might not know all the involved constraints here but I thought it would be worth mentioning anyway: Ibex in OpenTitan as well as our toolchains have support for the RISC-V Bitmanip extension ([see the Ibex documentation for more details](https://ibex-core.readthedocs.io/en/latest/03_reference/instruction_decode_execute.html#bit-manipulation-extension)), meaning you can leverage the `clmul(h)` and `grev` instructions for improving performance/code size of GHASH as outlined in [this paper (see Section 4)](https://tches.iacr.org/index.php/TCHES/article/view/8729/8329).
> 
> Thanks! I'm aware and plan to adjust the implementation to use the bitmanip instructions, but thought it would be helpful to get the simplest variant working first so that we have passing tests to start from

I agree, I think it's not a horrible idea to have both for off-target testing.

> I was also under the impression that the bitmanip instructions were not quite ready to use, but I guess I was mistaken. Was toolchain support added recently?

I can't comment on the maturity, e.g. does it only have the new instructions or can it optimize code to take advantage of the new instructions, but we have the updated toolchain with bimanip support thanks to @luismarques and @vogelpi since 01b2cace6.

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (alphan) [reply sw/device/lib/crypto/aes_gcm/aes_gcm.c]: I commented on the same line :) We may also consider moving this to `.bss`, i.e. making this a `static`.

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c (hunk 4 above)
Comment (alphan) [reply sw/device/lib/crypto/aes_gcm/aes_gcm.c]: Actually for the last chunk we can use:
```
memcpy(input_block.data, input, nbytes);
memset((char *)input_block.data + nbytes, 0, kAesBlockNumBytes - nbytes);
```

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (moidx) [reply sw/device/lib/crypto/aes_gcm/aes_gcm.c]: SGTM. I agree using 4bit is a good tradeoff for a memory versus performance perspective.

As a side note, I assume that if we end up implementing hardening on this, we'll choose the implementation that uses bitmanip instructions.

**Review by moidx — APPROVED** [review]

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c (hunk 10 above)
Comment (jadephilipoom) [reply sw/device/lib/crypto/aes_gcm/aes_gcm.c:116]: Yes, but it's much less readable. Now that I have a working version I'll give it another shot and see if it makes enough speed difference to be worthwhile!

**Comment by luismarques** [comment]
> Thanks for the clarification @jadephilipoom , your strategy makes complete sense. In addition, having a "simple" variant available will also allow us to evaluate the benefit of the bitmanip extension.
> 
> Bitmanip is ready to use in Ibex and OpenTitan - thanks @alphan for looking up the commit :-) I suspect you'll have to use intrinsics or inline assembly for `clmul(h)` and `grev` though. Maybe @luismarques know the details.

`grev*` is emitted automatically by Clang for some code patterns. But if you want tight control of the emitted instructions you'll probably want to use intrinsic or inline assembly anyway. For `clmul*` I think you always need intrinsics/assembly at the moment.

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c [hunk 15]
```diff
@@ -0,0 +1,356 @@
...
+ * @param H_table Precomputed product table for the hash subkey
+ * @param result Block in which to store output
+ */
+static void galois_mul(const aes_block_t *p, const aes_gcm_product_table_t *tbl,
+                       aes_block_t *result) {
+  // Initialize the result to 0.
+  memset(result->data, 0, kAesBlockNumBytes);
+
+  // 4-bit windows, so number of windows is 2x number of bytes.
+  size_t nwindows = kAesBlockNumBytes << 1;
```
Comment (dmcardle) [inline sw/device/lib/crypto/aes_gcm/aes_gcm.c:248]: Nit: Could this be `const size_t kNumWindows = ...`? I wish C had constexpr.

Comment (dmcardle) [reply sw/device/lib/crypto/aes_gcm/aes_gcm.c:248]: (My intent with this suggestion is to clarify on line 259 that that `nwindows` is not a parameter of `galois_mul()`.)

Comment (jadephilipoom) [reply sw/device/lib/crypto/aes_gcm/aes_gcm.c]: Done!
